pkg/utils/types/rwmaps.go

```diff
@@ -0,0 +1,73 @@
+package types
+
+import (
+	"sync"
+
+	"golang.org/x/exp/maps"
+)
+
+// RWMap is a map envelopment for concurrent map use
+type RWMap[K comparable, V any] struct {
+	m     map[K]V
+	mutex *sync.RWMutex
```

Comment:
IMO, We should follow object-oriented encapsulation principle, instead of exporting mutexes for external (to interface) users (and making a consumer responsbility for the object correctness). If, in the future, we want to change the object to have another synchronization mechanism, we would have to change all code that uses it (that is exactly what we are trying to avoid with encapsulation). The correct approach for the case you describe would be to provide a method that has a callback argument. The method would call the function while maintaining the safety of the context.

Reply:
Resuming, its either a "read-copy-update" approach (like the way its implemented NOW, where the user picks an item, works on it, and sets it back) OR a "transactional" approach (with atomicity, consistency, isolation) with a callback.

We have to think of the "types" being used in the map (its not the same depending on the types):

1. If the types are "simple" types (atomic uint, atomic int, pointer to a struct that has its internal synchronization) then its ok to be like that. Lets assume you're doing a map of an `example Struct{}`. You would get an instance of Example (one instantiated object) and rely on its own methods to change it (and it would have its own internal synchronization mechanism).

2. Now, if the types "complex" types (like a map of map, or map of array, non instantiated objects), then you are in trouble and you would require the callback approach, for example.

@yanivagman,

With that said, now you get why I proposed all the event definitions to be instantiated: to guarantee encapsulation, isolation and atomicity. By having everything instantiated we can always encapsulate synchronization mechanisms and make the users agnostic on how its done. I'm sorry @NDStrahilevitz did not like that as well, but the alternative to that model isn't great (like you can see in this simple example).

Reply:
So I agree with Rafael that I use this map to handle instanced objects, which are responsible to manage themselves.
We can have the map expose a callback function, which will be protected by `Lock()` for its entire operation.
@rafaeldtinoco maybe we should have 2 different types for it though, to make sure that the user is aware of what its use-case is and what best suit it.
I think that either way further PR can introduce the new callback method or the new type, so this conversation does not affect this PR.